File: src/cs/Ssh/Algorithms/KeyExchangeAlgorithm.cs
// Copyright (c) Microsoft. All rights reserved.

namespace Microsoft.DevTunnels.Ssh.Algorithms;

public abstract class KeyExchangeAlgorithm : SshAlgorithm
{
	protected KeyExchangeAlgorithm(
		string name,
		int keySizeInBits,
		string hashAlgorithmName,
		int hashDigestLength)
		: base(name)
	{
		this.KeySizeInBits = keySizeInBits;
		this.HashAlgorithmName = hashAlgorithmName;
		this.HashDigestLength = hashDigestLength;
	}

	public int KeySizeInBits { get; }
	public string HashAlgorithmName { get; }
	public int HashDigestLength { get; }

	public abstract IKeyExchange CreateKeyExchange();
}

File: src/ts/ssh/algorithms/keyExchangeAlgorithm.ts
//
//  Copyright (c) Microsoft Corporation. All rights reserved.
//

import { Buffer } from 'buffer';
import { SshAlgorithm } from './sshAlgorithm';
import { Signer } from './hmacAlgorithm';

export abstract class KeyExchangeAlgorithm implements SshAlgorithm {
	protected constructor(
		public readonly name: string,
		public readonly keySizeInBits: number,
		public readonly hashAlgorithmName: string,
		public readonly hashDigestLength: number,
	) {}

	public abstract createKeyExchange(): KeyExchange;
}

export interface KeyExchange extends Signer {
	startKeyExchange(): Promise<Buffer>;
	decryptKeyExchange(exchangeValue: Buffer): Promise<Buffer>;
}
